### PR TITLE
support maintaining WebView and iFrame state when moving WebView in view hierarchy

### PR DESCRIPTION
**Pull request by donaldchen**
By default, moving a React `WebView` component in the view hierarchy creates a new native WebView and refreshes the iframe. We need the ability to move the `WebView` without refreshing the iframe, and this PR adds support for that by reusing the native WebView instance.

**High level react-native-webview API changes**
This PR updates the React Native JS `WebView` API to have two additional props:
- `keepWebViewInstanceAfterUnmount` - a boolean that decides whether the WebView should get released when the react component unmounts
- `webViewKey` - a key for accessing specific WebView instances

**How This Works**

For iOS:
We have a singleton map where the keys are strings, and the values are `WKWebViews`. When React Native asks for a new view, we create a new `RNCWebView` instance. `RNCWebView` is a parent of `WKWebView`, and when the props say to reuse the native WebView instance, `RNCWebView` looks for the existing `WKWebView` instance from the map before it decides whether it needs to create a new `WKWebView` instance. Additionally, when we move a `WKWebView` between two `RNCWebView` parents, we need to remove the reference to the old parent that was held through the observer. We also have a map of web view keys  to `RNCWebView`s, so we can remove the correct observer from the `WKWebView`. If we don't remove the observer when we change the `WKWebView`'s parent, the app will crash with an `EXC_BAD_ACCESS` error.

**Before and After**
Additionally, we're adding a public function on `WebView` called `release` for feature code to explicitly release the native WebView instance when it’s not getting released on unmount.

Here's the behavior where the iframe refreshes when `keepWebViewInstanceAfterUnmount` is false or undefined.

https://user-images.githubusercontent.com/3821698/165408230-849f75bc-771d-4658-8171-5b64fb1bf277.mov

Here's the behavior where the iframe maintains its state as the react WebView moves in the view hierarchy when `keepWebViewInstanceAfterUnmount` is true and `webViewKey` is defined.

https://user-images.githubusercontent.com/3821698/166070390-78b6e6be-701b-4e03-9bd2-b5c9c57cda17.mov

### CODE REVIEW

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 1]
```diff
@@ -139,6 +139,9 @@ public class RNCWebViewManager extends SimpleViewManager<WebView> {
   public static final int COMMAND_LOAD_URL = 7;
   public static final int COMMAND_FOCUS = 8;
 
+  // commands added by Discord
+  public static final int COMMAND_RELEASE = 4001;
```
Comment (donaldchen) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java:143]: I'm starting this from 4001, so there will be fewer collisions with other commands in the future. Later, I may make a PR for these changes into the non-discord react-native-webview repo. If I'll do that, I'll change this value to `9`.

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 2]
```diff
@@ -754,6 +758,9 @@ public void receiveCommand(WebView root, int commandId, @Nullable ReadableArray
       case COMMAND_CLEAR_HISTORY:
         root.clearHistory();
         break;
+      case COMMAND_RELEASE:
+        // no-op for now
```
Comment (donaldchen) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java:762]: The iOS code in this PR is ready for review, and while that's getting reviewed, I'll see if I can get an Android implementation working.

**Review by nealmanaktola — APPROVED** [review]
looks good, did a fairly quick pass, but didn't see any issues

**Comment by donaldchen** [comment]
Thanks for the review! I'm going to recreate this PR, since I realized I messed up the branches. This branch should be named  `ios-reuse-web-view-instance`, and this should merge into a branch called `11.18.1-discord` rather than into `master`.

**Pull request merged**